### ISSUE TITLE
acceptance: local cluster is blocked on problems with empty join flag (#13027)

### ISSUE THREAD

**Issue by tschottdorf** [cockroachdb/cockroach#18027]
See https://github.com/cockroachdb/cockroach/pull/18007.

`TestMixedVersion` runs with three nodes as otherwise it's flaky: the node first restarted (i.e. the one without join flags) never manages to connect to gossip. See the comment within that test.

One theory is that the other three nodes manage to connect among each other sufficiently so that the don't feel that establishing new connections is helpful. In turn, the first node remains alone and unable to do anything about it. The theory is easy to test by running the test with a version of cockroach that always connects to everything it can.

It repros easily using `time go test -v -tags acceptance ./pkg/acceptance/ -run Mixed  -show-logs -count 100 -timeout 1h` within only a few iters (after making the test 4nodes, of course).

**Comment by bdarnell** [comment]
> the one without join flags

It's started without join flags the first time, but on subsequent restarts the recommended practice would be to add the join flag. Are you doing that here?

Are the restarted nodes coming back to the same ports they used the first time or are they using port zero again? In that case I think your hypothesis about the other three nodes forming connections between themselves is plausible.

**Comment by tschottdorf** [comment]
The nodes are restarted with new, random ports. We add all the join flags we can, as in: the i-th started node gets i-1 addresses to join to, and nodes are started in parallel (i.e. we only wait for the port to be known, then fire up the next one).

**Comment by bdarnell** [comment]
I see. So you can't give the first node to restart a join flag even if you wanted to. I think giving all the other nodes *only* the first node's address would work (then they couldn't form a connected network on their own that didn't include the first node). 

Someone was recently talking about adding a way to dynamically update a node's join flag, which would be useful here (maybe this issue is where it came from).

**Comment by tschottdorf** [comment]
Yeah, I was suggesting that. You're right that pointing everyone at the first node would do the trick. I'll try that.

**Comment by tschottdorf** [comment]
I've implemented this, but unfortunately I'm still seeing a deadlock in `TestGossipRestart`:

```
I170830 14:35:25.092692 1121 acceptance/localcluster/cluster.go:604  process 12749 starting: [/Users/tschottdorf/go/src/github.com/cockroachdb/cockroach/cockroach start --insecure --host=127.0.0.1 --port=0 --http-port=0 --store=/Users/tschottdorf/go/src/github.com/cockroachdb/cockroach/pkg/acceptance/.localcluster518785475/1 --listening-url-file=/Users/tschottdorf/go/src/github.com/cockroachdb/cockroach/pkg/acceptance/.localcluster518785475/1/cockroachdb-url --cache=256MiB]
I170830 14:35:25.290655 1121 acceptance/localcluster/cluster.go:604  process 12750 starting: [/Users/tschottdorf/go/src/github.com/cockroachdb/cockroach/cockroach start --insecure --host=127.0.0.1 --port=0 --http-port=0 --store=/Users/tschottdorf/go/src/github.com/cockroachdb/cockroach/pkg/acceptance/.localcluster518785475/2 --listening-url-file=/Users/tschottdorf/go/src/github.com/cockroachdb/cockroach/pkg/acceptance/.localcluster518785475/2/cockroachdb-url --cache=256MiB --join 127.0.0.1:54882]
I170830 14:35:25.470710 1121 acceptance/localcluster/cluster.go:604  process 12751 starting: [/Users/tschottdorf/go/src/github.com/cockroachdb/cockroach/cockroach start --insecure --host=127.0.0.1 --port=0 --http-port=0 --store=/Users/tschottdorf/go/src/github.com/cockroachdb/cockroach/pkg/acceptance/.localcluster518785475/3 --listening-url-file=/Users/tschottdorf/go/src/github.com/cockroachdb/cockroach/pkg/acceptance/.localcluster518785475/3/cockroachdb-url --cache=256MiB --join 127.0.0.1:54882]
I170830 14:35:25.658336 1121 acceptance/localcluster/cluster.go:604  process 12752 starting: [/Users/tschottdorf/go/src/github.com/cockroachdb/cockroach/cockroach start --insecure --host=127.0.0.1 --port=0 --http-port=0 --store=/Users/tschottdorf/go/src/github.com/cockroachdb/cockroach/pkg/acceptance/.localcluster518785475/4 --listening-url-file=/Users/tschottdorf/go/src/github.com/cockroachdb/cockroach/pkg/acceptance/.localcluster518785475/4/cockroachdb-url --cache=256MiB --join 127.0.0.1:54882]
```

All nodes restarted pointing at the same node, but none of the nodes actually gets to serving. Looking at the logs, all but the first one claim to have connected to gossip. Perhaps the first range is unavailable (caught in a middle of a rebalance? Does that explain it or should we survive with one node gone?), but all of the debug tooling breaks, including problem ranges (cc @BramGruneir):

![image](https://user-images.githubusercontent.com/5076964/29889372-f27e3d80-8d91-11e7-91a6-c34f2a3be2be.png)

**Comment by tschottdorf** [comment]
(I should clarify that the first range is definitely unavailable and hasn't gossiped the first range descriptor in the last posting).

**Comment by bdarnell** [comment]
Try running with COCKROACH_ENABLE_PREVOTE=false (inspired by #17989)

**Comment by tschottdorf** [comment]
Manually restarting the first node with a reasonable join flag brought it into the state of the other nodes: gossip-connected, but otherwise broken. Interestingly, the range1 page now loads and shows (on n1)

![image](https://user-images.githubusercontent.com/5076964/29889772-5f3bb294-8d93-11e7-9cc6-674d691229d9.png)

Unfortunately, the `#/reports/network` endpoint is useless:

![image](https://user-images.githubusercontent.com/5076964/29889823-86d52498-8d93-11e7-9ec9-e9ce895696ea.png)

Still, something seems to be going on with not gossiping the relevant NodeDescriptors perhaps? On Node 2, we see only an empty page for range1 ("maybe it doesn't exist?"). Looking at http://node2/_status/gossip/local confirms that it has only itself, the cluster ID, and, randomly, the distsql version.

@bdarnell does the reinforce your prevote suspicion? Even if this *is* PreVote though, does it explain the missing gossip'ed node addresses above? Node2 shows a completely empty node list in the ui. `./cockroach debug gossip-values` of course times out.

![image](https://user-images.githubusercontent.com/5076964/29889851-9fbad390-8d93-11e7-9421-62b1e99310c1.png)

**Comment by tschottdorf** [comment]
Confirmed (looking at all gossip endpoints) that node1 and node4 have exchanged their addresses, and that n2 and n3 haven't. I doubt we can blame this on prevote -- somehow the descriptors got ignored.

Ah, fantastic. I think I have an idea of what's happening here. The node descriptor is gossiped periodically in `startGossip`. `startGossip` also triggers gossip of the first range. Unfortunately, this will block on the lease on that range. If not enough nodes gossip their address early, this can deadlock. I think a majority of the replicas of the first range are stuck as below, and since they are stuck, they will never get around to updating their node descriptor, so they will never get to unclog the first range since no(t enough) Raft traffic is possible.

That sounds like something that we can fix using a timeouted context.

```
goroutine 113 [select, 36 minutes]:
github.com/cockroachdb/cockroach/pkg/storage.(*Replica).redirectOnOrAcquireLease.func2(0xc42086f8f8, 0xc42059c8a0, 0xc42035a000, 0x6301900, 0xc420680780, 0xc420692000, 0x0)
	/Users/tschottdorf/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:1350 +0x228
github.com/cockroachdb/cockroach/pkg/storage.(*Replica).redirectOnOrAcquireLease(0xc42035a000, 0x6301900, 0xc420680780, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/tschottdorf/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:1401 +0x217
github.com/cockroachdb/cockroach/pkg/storage.(*Replica).getLeaseForGossip.func1(0x6301900, 0xc420680780)
	/Users/tschottdorf/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:4960 +0x6e
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunTask(0xc42038a480, 0x6301900, 0xc420680780, 0x57551b4, 0x2a, 0xc420042b48, 0x0, 0x0)
	/Users/tschottdorf/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:230 +0xf7
github.com/cockroachdb/cockroach/pkg/storage.(*Replica).getLeaseForGossip(0xc42035a000, 0x6301900, 0xc420680780, 0xc420042c18, 0x10)
	/Users/tschottdorf/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:4956 +0x123
github.com/cockroachdb/cockroach/pkg/storage.(*Replica).maybeGossipFirstRange(0xc42035a000, 0x6301900, 0xc420680780, 0xc420680780)
	/Users/tschottdorf/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:5018 +0x20e
github.com/cockroachdb/cockroach/pkg/storage.(*Store).startGossip.func3(0x6301900, 0xc420680780, 0xc42035a000, 0x6301900, 0xc420680780)
	/Users/tschottdorf/go/src/github.com/cockroachdb/cockroach/pkg/storage/store.go:1241 +0x3f
github.com/cockroachdb/cockroach/pkg/storage.(*Store).startGossip.func4(0x6301900, 0xc420680750)
	/Users/tschottdorf/go/src/github.com/cockroachdb/cockroach/pkg/storage/store.go:1278 +0x19b
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker.func1(0xc4201ffc60, 0xc42038a480, 0xc4203fe460)
	/Users/tschottdorf/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:197 +0xf3
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker
	/Users/tschottdorf/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:190 +0xad
```

**Comment by tschottdorf** [comment]
Hrm, I looked at the wrong goroutine (I wanted `(*Node).startGossip` to hang). Time to sleuth some more...

**Comment by tschottdorf** [comment]
Ah. The descriptor is gossiped at one hour frequency:

```
	// gossipNodeDescriptorInterval is the interval for gossiping the node descriptor.
	// Note that increasing this duration may increase the likelihood of gossip
	// thrashing, since node descriptors are used to determine the number of gossip
	// hops between nodes (see #9819 for context).
	gossipNodeDescriptorInterval = 1 * time.Hour
```

First of all that seems pretty high, so lowering this will probably "fix" things (if gossip takes the value as "new" and actually broadcasts again).

But, why didn't gossip distribute the first value? Each node definitely has their own descriptor in gossip, and presumably they would distribute them.

Oh wait:

```
$ grep -h 'gossip status' /Users/tschottdorf/go/src/github.com/cockroachdb/cockroach/pkg/acceptance/.localcluster518785475/*/logs/cockroach.log | sort | tail -n 5
I170830 15:22:27.076093 106 gossip/gossip.go:459  [n4] gossip status (stalled, 2 nodes)
I170830 15:23:08.486667 120 gossip/gossip.go:459  [n1] gossip status (stalled, 2 nodes)
I170830 15:23:26.638037 119 gossip/gossip.go:459  [n2] gossip status (stalled, 1 node)
I170830 15:23:26.821893 119 gossip/gossip.go:459  [n3] gossip status (stalled, 1 node)
I170830 15:23:27.077588 106 gossip/gossip.go:459  [n4] gossip status (stalled, 2 nodes)
```

duh. Of course, by restarting node1 I robbed everyone else of their one `--join` flag. Apparently they hadn't connected to one another prior to that. Wonder why not.

**Comment by tschottdorf** [comment]
I restarted the first node on its previous port (54882) and things didn't get much better. I just realized that nodes 2 and 3 aren't running any more. I'm not sure why and there's nothing in the logs. Perhaps I accidentally killed them, though how I'm not sure.

**Comment by tschottdorf** [comment]
With three nodes, `TestGossipRestart` passes 20 consecutive runs locally. I'm not sure what's up here, but I'll restrict that test to three nodes to get my change in. @a-robinson, do you have any conjectures handy?

**Comment by tschottdorf** [comment]
New run, new problem: the first node is stuck again. Interestingly it's stuck here:

```
goroutine 38 [select, 3 minutes]:
github.com/cockroachdb/cockroach/pkg/server.(*Node).connectGossip(0xc4205eef00, 0x6307900, 0xc4206f00f0, 0x1, 0x300000000)
	/Users/tschottdorf/go/src/github.com/cockroachdb/cockroach/pkg/server/node.go:607 +0x1d3
github.com/cockroachdb/cockroach/pkg/server.(*Node).startStores(0xc4205eef00, 0x6307900, 0xc4206f00f0, 0xc42049c0f8, 0x1, 0x1, 0xc420011200, 0x0, 0x0)
	/Users/tschottdorf/go/src/github.com/cockroachdb/cockroach/pkg/server/node.go:506 +0x6bc
github.com/cockroachdb/cockroach/pkg/server.(*Node).start(0xc4205eef00, 0x6307900, 0xc4206f00f0, 0x62f7440, 0xc4202860c0, 0xc42020dc20, 0x1, 0x1, 0x0, 0x0, ...)
	/Users/tschottdorf/go/src/github.com/cockroachdb/cockroach/pkg/server/node.go:383 +0x32d
github.com/cockroachdb/cockroach/pkg/server.(*Server).Start(0xc42009af00, 0x6307900, 0xc4206f00f0, 0xf, 0x0)
	/Users/tschottdorf/go/src/github.com/cockroachdb/cockroach/pkg/server/server.go:855 +0x1730
github.com/cockroachdb/cockroach/pkg/cli.runStart.func1.2(0x6307900, 0xc42030eea0, 0xc420312200, 0xc420011200, 0xc420308020, 0xbe61e55b50730628, 0x26ecf3d, 0x67fe720, 0x0, 0x0)
	/Users/tschottdorf/go/src/github.com/cockroachdb/cockroach/pkg/cli/start.go:363 +0x1b6
github.com/cockroachdb/cockroach/pkg/cli.runStart.func1(0xc420312200, 0x6307900, 0xc42030eea0, 0x6314c00, 0xc420240dc0, 0xc420011200, 0xc420308020, 0xbe61e55b50730628, 0x26ecf3d, 0x67fe720, ...)
	/Users/tschottdorf/go/src/github.com/cockroachdb/cockroach/pkg/cli/start.go:428 +0x11c
created by github.com/cockroachdb/cockroach/pkg/cli.runStart
	/Users/tschottdorf/go/src/github.com/cockroachdb/cockroach/pkg/cli/start.go:330 +0x44c
```

This is notably before the point at which we start serving the cmux listener:

```go
	// Now that we have a monotonic HLC wrt previous incarnations of the process,
	// init all the replicas. At this point *some* store has been bootstrapped or
	// we're joining an existing cluster for the first time.
	err = s.node.start(
		ctx,
		unresolvedAdvertAddr,
		bootstrappedEngines, emptyEngines,
		s.cfg.NodeAttributes,
		s.cfg.Locality,
		cv,
	)

        // ...

	log.Infof(ctx, "starting %s server at %s", s.cfg.HTTPRequestScheme(), unresolvedHTTPAddr)
	log.Infof(ctx, "starting grpc/postgres server at %s", unresolvedListenAddr)
	log.Infof(ctx, "advertising CockroachDB node at %s", unresolvedAdvertAddr)
	s.stopper.RunWorker(workersCtx, func(context.Context) {
		serveOnMux.Do(func() {
			netutil.FatalIfUnexpected(m.Serve())
		})
	})
```

This raises the interesting question of: how does this ever work? And how are the other nodes claiming that they're connected to gossip? Are we perhaps persisting something and the other nodes immediately think they're golden?

**Comment by tschottdorf** [comment]
Yep, looks like the range on the first lease is still active on the other nodes (and the first range tends to be on all nodes but the first, presumably it gets rebalanced first):

```
goroutine 107 [running]:
runtime/debug.Stack(0x1, 0xc400000002, 0x57249b2)
        /usr/local/Cellar/go/1.9/libexec/src/runtime/debug/stack.go:24 +0xa7
runtime/debug.PrintStack()
        /usr/local/Cellar/go/1.9/libexec/src/runtime/debug/stack.go:16 +0x22
github.com/cockroachdb/cockroach/pkg/gossip.(*Gossip).signalConnectedLocked(0xc420064c00)
        /Users/tschottdorf/go/src/github.com/cockroachdb/cockroach/pkg/gossip/gossip.go:1270 +0xb8
github.com/cockroachdb/cockroach/pkg/gossip.(*Gossip).addInfoLocked(0xc420064c00, 0x5721d35, 0xa, 0xc4209efc18, 0x10, 0x10, 0x0, 0x1, 0x0)
        /Users/tschottdorf/go/src/github.com/cockroachdb/cockroach/pkg/gossip/gossip.go:794 +0xd5
github.com/cockroachdb/cockroach/pkg/gossip.(*Gossip).AddInfo(0xc420064c00, 0x5721d35, 0xa, 0xc4209efc18, 0x10, 0x10, 0x0, 0x0, 0x0)
        /Users/tschottdorf/go/src/github.com/cockroachdb/cockroach/pkg/gossip/gossip.go:786 +0xcf
github.com/cockroachdb/cockroach/pkg/storage.(*Replica).maybeGossipFirstRange(0xc4201bf400, 0x6307900, 0xc4206bf290, 0xc4206bf290)
        /Users/tschottdorf/go/src/github.com/cockroachdb/cockroach/pkg/storage/replica.go:5034 +0x13d
github.com/cockroachdb/cockroach/pkg/storage.(*Store).startGossip.func3(0x6307900, 0xc4206bf290, 0xc4201bf400, 0x6307900, 0xc4206bf290)
        /Users/tschottdorf/go/src/github.com/cockroachdb/cockroach/pkg/storage/store.go:1242 +0x3f
github.com/cockroachdb/cockroach/pkg/storage.(*Store).startGossip.func4(0x6307900, 0xc4206bf260)
        /Users/tschottdorf/go/src/github.com/cockroachdb/cockroach/pkg/storage/store.go:1279 +0x19b
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker.func1(0xc4202703c0, 0xc42071c000, 0xc42036ff90)
        /Users/tschottdorf/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:198 +0xf3
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker
        /Users/tschottdorf/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:191 +0xad
```

So you have three nodes that do their own thing, but the first node out of its own volition can't ever come back to life.

I just added a 15s sleep between the cluster kill and the cluster restart, and the test still sometimes passes. When it does, I see that the first node gossips itself to life because it has a replica of the first range (that's enough to gossip the cluster ident, no lease required).

I'm going to go out on a limb and say that the cases in which the deadlock occurs are precisely when at the time of restart, node1 does not have a replica of the first range. In that case, the first node will not open its gossip port. It can't talk to anyone else, and nobody can talk to it. Deadlock!

**Comment by tschottdorf** [comment]
Yeah, I think that's on the money. The test just passed a dozen times with this diff that simply opens the listener early.

It's probably not something that can merge without a healthy dose of scrutiny and archaeology, but at least we know what's up.

```diff
diff --git a/pkg/server/server.go b/pkg/server/server.go
index c5ff6d6d5..1dff72404 100644
--- a/pkg/server/server.go
+++ b/pkg/server/server.go
@@ -842,6 +842,12 @@ func (s *Server) Start(ctx context.Context) error {
                atomic.StoreInt32(&initLActive, 0)
        }

+       s.stopper.RunWorker(workersCtx, func(context.Context) {
+               serveOnMux.Do(func() {
+                       netutil.FatalIfUnexpected(m.Serve())
+               })
+       })
+
        // We ran this before, but might've bootstrapped in the meantime. This time
        // we'll get the actual list of bootstrapped and empty engines.
        bootstrappedEngines, emptyEngines, cv, err := inspectEngines(ctx, s.engines, s.cfg.Settings.Version.MinSupportedVersion, s.cfg.Settings.Version.ServerVersion)
@@ -916,11 +922,6 @@ func (s *Server) Start(ctx context.Context) error {
        log.Infof(ctx, "starting %s server at %s", s.cfg.HTTPRequestScheme(), unresolvedHTTPAddr)
        log.Infof(ctx, "starting grpc/postgres server at %s", unresolvedListenAddr)
        log.Infof(ctx, "advertising CockroachDB node at %s", unresolvedAdvertAddr)
-       s.stopper.RunWorker(workersCtx, func(context.Context) {
-               serveOnMux.Do(func() {
-                       netutil.FatalIfUnexpected(m.Serve())
-               })
-       })

        if len(s.cfg.SocketFile) != 0 {
                log.Infof(ctx, "starting postgres server at unix:%s", s.cfg.SocketFile)
```

**Comment by tschottdorf** [comment]
A lengthy blame chase points at https://github.com/cockroachdb/cockroach/pull/5441 as the primary motivator in moving the listener so far down.

If we can refactor things so that we open ourselves up to gossip but postpone the rest, that would be good. Or not postpone the rest, but it seems that we do want to avoid RPCs to reach server components before they're fully set up.

Summoning @cuongdo for triage.

**Comment by a-robinson** [comment]
Sorry for not helping out on this earlier! It looks like you independently discovered the same problem as https://github.com/cockroachdb/cockroach/issues/13027, which is also [documented in our known limitations](https://www.cockroachlabs.com/docs/stable/known-limitations.html#join-flags-when-restoring-a-backup-onto-new-machines) (albeit with a focus on restore). That issue has a 1.1 milestone, but didn't get any love until now since it's not something people come across often.

If node addresses change across restarts, join flags currently have to form a fully-connected directed graph. This isn't an issue in the normal case when node addresses don't change because each node should have already persisted enough bootstrap addresses to form such a graph.

**Comment by tschottdorf** [comment]
I should add that while we're at it, it's also fairly counterintuitive that nodes who can't talk to any of their peers declare themselves "connected to gossip" just because they have a replica of the first range. I *think* morally what we mean by "gossip is connected" is that we see a quorum of peers that have the sentinel gossip key. I don't think we're very likely to implement that notion, but perhaps we should clarify the commentary surrounding "connectedness".

@a-robinson ah, thanks for the pointer. That's a shame, but it looks like we can fix it.

**Comment by petermattis** [comment]
The work on `init` showed us how to listen on only a specific service initially and then add more services later. Can we do something similar and listen on gossip immediately?

**Comment by tschottdorf** [comment]
@petermattis I'll take a look, maybe this is straightforward.

**Comment by petermattis** [comment]
@tschottdorf For reference, see `Server.Start` and how `initL` is defined. In that case, `initL` intercepts all connections during initialization. We'd want something a bit different for gossip, but it seems feasible.

**Comment by tschottdorf** [comment]
I don't think it's quite that easy. The init listener becomes the only thing that listens on the node and later deactivates itself, at which point other connections can be made. We are in the situation in which we want to serve grpc traffic, but only for certain endpoints. I could set it up so that we serve first gossip RPCs only (i.e. you can talk gRPC to the port but no endpoint will exist except gossip) and later also allow other ones, but that's not the point - we want our gRPC connections to hang until the listener becomes active, and then when we want all the endpoints to be registered already. That would require something like cmux, but peeking inside of the grpc protocol.

Still possible (accept connections, wait until we're out of "gossip only mode", and then pass them back into the new grpc server), but tricky. How many connections do we accept? What once we have too many?

The option of opening Gossip early and "404" everything else (i.e. initL-like behavior) still makes sense to me. The basic question is whether it's acceptable to fail other RPCs for some time. I think it's perfectly reasonable, just not what we're doing so far. We have `--listening-file-url` and health HTTP endpoints (which we can continue to open late) for load balancers so that they know when we're good to go.

**Comment by bdarnell** [comment]
#5441 was trying to fix synchronization in tests. We have better ways of doing that now. I think it ought to be fine for a node with at least one initialized store to just start up all its listeners/servers early instead of waiting for gossip. I don't think we need to "404" the other kinds of RPCs, we just need to make sure that they return retryable errors while in this state. (I *think* the error we'd be concerned with here is the DistSender's `firstRangeMissingError`. We should probably sleep and retry that one a few times before returning it).

**Comment by tschottdorf** [comment]
I just hacked together the `initL` approach. As in,

```
./cockroach node ls --insecure
Error: rpc error: code = Unimplemented desc = unknown service cockroach.server.serverpb.Status
Failed running "node"
```

Then later it works. If we can't encapsulate it at the `(*Server).Start` level, then opening the listeners early and hoping that every endpoint does the right thing during startup sounds awkward to me. Perhaps we can just handle the above error where it becomes relevant (`DistSender`, ...) as retryable and non-ambiguous?

**Comment by tschottdorf** [comment]
Maybe this is a silly idea, but can we use a grpc interceptor which we use for all connections that we're serving? It won't do anything but delay all the non-gossip RPCs until we flip the "initialized" switch. Or it just straight-up returns an error of our choosing.

**Comment by tschottdorf** [comment]
Explored the interceptor idea (the obvious problem is that it adds a little bit of per-RPC overhead). A problem with the other one (initL-style gossip initialization) is that we're starting a separate grpc server, and there's no good way of shutting it down -- we'd either have to forcibly disconnect gossip clients (which is fine but causes annoying log messages) or do a graceful stop and live with that never actually happening due to the long-lived gossip connections. Or we have to instruct gossip to shed all clients. Either way, there's complexity there that doesn't seem worth it.

```
$ ./cockroach node ls --insecure
Error: rpc error: code = Unknown desc = not serving /cockroach.server.serverpb.Status/Nodes yet
Failed running "node"
$ ./cockroach node ls --insecure
+----+
| id |
+----+
|  1 |
+----+
(1 row)
```

I'm reasonably optimistic that we can return any grpc error code that we want.

```diff
diff --git a/pkg/rpc/context.go b/pkg/rpc/context.go
index 853dbb3dc..6d0dd714a 100644
--- a/pkg/rpc/context.go
+++ b/pkg/rpc/context.go
@@ -15,7 +15,6 @@
 package rpc
 
 import (
-	"errors"
 	"fmt"
 	"math"
 	"net"
@@ -26,6 +25,7 @@ import (
 
 	"github.com/grpc-ecosystem/grpc-opentracing/go/otgrpc"
 	opentracing "github.com/opentracing/opentracing-go"
+	"github.com/pkg/errors"
 	"github.com/rubyist/circuitbreaker"
 	"golang.org/x/net/context"
 	"golang.org/x/sync/syncmap"
@@ -133,15 +133,40 @@ func NewServer(ctx *Context) *grpc.Server {
 		}
 		opts = append(opts, grpc.Creds(credentials.NewTLS(tlsConfig)))
 	}
+
+	var tracingInterceptor grpc.UnaryServerInterceptor
 	if tracer := ctx.AmbientCtx.Tracer; tracer != nil {
 		// We use a SpanInclusionFunc to save a bit of unnecessary work when
 		// tracing is disabled.
-		interceptor := otgrpc.OpenTracingServerInterceptor(
+		tracingInterceptor = otgrpc.OpenTracingServerInterceptor(
 			tracer,
 			otgrpc.IncludingSpans(otgrpc.SpanInclusionFunc(spanInclusionFunc)),
 		)
-		opts = append(opts, grpc.UnaryInterceptor(interceptor))
 	}
+
+	hack := make(chan struct{})
+	go func() {
+		time.Sleep(10 * time.Second)
+		log.Info(context.TODO(), "now serving RPCs")
+		close(hack)
+	}()
+
+	interceptor := func(
+		ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler,
+	) (interface{}, error) {
+		select {
+		case <-hack:
+		default:
+			return nil, errors.Errorf("not serving %s yet", info.FullMethod)
+		}
+		if tracingInterceptor == nil {
+			return handler(ctx, req)
+		}
+		return tracingInterceptor(ctx, req, info, handler)
+	}
+
+	opts = append(opts, grpc.UnaryInterceptor(interceptor))
+
 	s := grpc.NewServer(opts...)
 	RegisterHeartbeatServer(s, &HeartbeatService{
 		clock:              ctx.LocalClock,

```

**Comment by bdarnell** [comment]
I don't think it would be too bad to start the servers early and make sure it works with disconnected gossip (we already have to support the case when gossip becomes disconnected later). But if you're nervous about that an interceptor could work too (and I agree that that's better than a separate listener).

**Comment by a-robinson** [comment]
I wouldn't feel very safe about running before gossip is connected until we have https://github.com/cockroachdb/cockroach/issues/18058 in place to make sure the node we're talking to is from the same cluster and at a valid version.

**Comment by tschottdorf** [comment]
I saw TestDecommission time out during a restart, https://teamcity.cockroachdb.com/viewLog.html?buildId=376299&buildTypeId=Cockroach_UnitTests

Doesn't seem to happen often, but something similar might've happened there.

**Comment by tschottdorf** [comment]
As a follow-up to the above, random hangs are still happening. When they do, it's usually during a node restart. In the example I'm looking at right now, the fourth node has an active gossip connection to the first node but remains stalled, and generally all nodes are unhappy with all of their heartbeats timing out. The gossip connectivity graph snagged from the `gossip status:` logs is

```
1 <-> 2
3 <-> 4
```

Womp womp. By the way, I see this on my own PRs once or twice a week, so it's happening quite frequently. On restart, the node always tries to join only to the "oldest up node", but since we're doing a rolling restart in random order, this can result. In this example, we restarted the nodes in this order: 1 3 1 2 1 4. Reading from back to front, this explains why 4 got joined to 3 (it had been up the longest at that point). Before that, 1 got started and told to join 3 as well, but apparently under an incorrect port (bug in the test framework, apparently -- node likely hadn't updated the port yet) -- and then it used node2, the (still valid) address of which it had persisted. Without gossip persistence, this node would've been unable to connect at all (I think that's generally preferable, and would result if we implement the "changed join flags means clearing persistence" heuristic which would've also given that outcome).

These issues are pretty annoying and in dynamically provisioned environments they're not as infrequent as we'd like. I have no immediate idea how to solve them, though.

**Comment by tschottdorf** [comment]
> (bug in the test framework, apparently -- node likely hadn't updated the port yet)

Ah, no, no such bug. When 1 restarts the last time, 4 is actually the longest-running node, and so 3 joins to it (but then 4 is killed next).

To finish up the thread, when 2 restarts it also gets 4 (which later restarts), etc. It's not predictable which config you end up in. We need to be able to recover from such a split.
